shark_turbine/kernel/_support/nodes.py

```diff
@@ -0,0 +1,151 @@
+from abc import ABC
+from dataclasses import dataclass
+from typing import Any, Optional, Sequence, Type, TypedDict, Union
+import torch.fx as fx
+
+from ..lang.functional_types import Memory, Register
+from .._support.indexing import IndexExpr
+from .._support.dtype import DataType
+
+
+def get_node_name(string: str, skip_first: bool = True):
+    snakeString = ""
+    if skip_first:
+        snakeString += string[0].lower()
+        string = string[1:]
+    for i in string:
+        if i.isupper():
+            snakeString += "_" + i.lower()
+        else:
+            snakeString += i
+    # Drop the "_node" suffix
+    return snakeString[:-5]
+
+
+@dataclass
+class CustomNode(ABC):
+    """
+    Base class for all custom fx nodes.
+    """
+
+    op: Any
+
+    @classmethod
+    def from_fx_node(cls, node: fx.Node):
+        instance = cls(node.graph, node.op, *node.args)
+        instance.fx_node = node
+        return instance
+
+    def __str__(self) -> str:
+        name = get_node_name(self.__class__.__name__)
+        vars_list = [f"{key}={value}" for key, value in vars(self).items()][2:]
+        vars_str = ", ".join(vars_list)
+        return f"{name} {vars_str}" + "\n"
+
+    def custom_string(self, value_map: dict[str, str]) -> str:
+        # If a subclass does not define custom printing we revert to the default
+        return str(self)
+
+    def add_to_graph(self, graph: fx.Graph):
```

Comment:
This function is a bit confusing. When adding a node to a graph, I would have expected a graph.create_node. Instead there is a graph.create_proxy here. What is the difference? Maybe consider calling this add_proxy?